frugen.c

```diff
@@ -816,120 +923,245 @@ int main(int argc, char *argv[])
+			json_object_add_with_type(section, "serial", chassis.serial.val, chassis.serial.type);
+			temp_obj = json_object_new_array();
+			fru_reclist_t *next = chassis.cust;
+			while (next != NULL) {
+				json_object_add_with_type(temp_obj, NULL, next->rec->data,
+							  typelen2ind(next->rec->typelen));
+				next = next->next;
+			}
+			json_object_object_add(section, "custom", temp_obj);
+			json_object_object_add(json_root, "chassis", section);
+		}
 
-	fname = argv[optind];
-	debug(1, "FRU info data will be stored in %s", fname);
+		if (has_product) {
+			section = json_object_new_object();
+			temp_obj = json_object_new_int(product.lang);
+			json_object_object_add(section, "lang", temp_obj);
+			json_object_add_with_type(section, "mfg", product.mfg.val, product.mfg.type);
+			json_object_add_with_type(section, "pname", product.pname.val, product.pname.type);
+			json_object_add_with_type(section, "serial", product.serial.val, product.serial.type);
+			json_object_add_with_type(section, "pn", product.pn.val, product.pn.type);
+			json_object_add_with_type(section, "ver", product.ver.val, product.ver.type);
+			json_object_add_with_type(section, "atag", product.atag.val, product.atag.type);
+			json_object_add_with_type(section, "file", product.file.val, product.file.type);
+			temp_obj = json_object_new_array();
+			fru_reclist_t *next = product.cust;
+			while (next != NULL) {
+				json_object_add_with_type(temp_obj, NULL, next->rec->data,
+							  typelen2ind(next->rec->typelen));
+				next = next->next;
+			}
+			json_object_object_add(section, "custom", temp_obj);
+			json_object_object_add(json_root, "product", section);
+		}
 
-	if (has_internal) {
-		debug(1, "FRU file will have an internal use area");
-		/* Nothing to do here, added for uniformity, the actual
-		 * internal use area can only be initialized from file */
-	}
+		if (has_board) {
+			section = json_object_new_object();
+			temp_obj = json_object_new_int(board.lang);
+			json_object_object_add(section, "lang", temp_obj);
+			json_object_add_with_type(section, "time", timebuf, 0);
+			json_object_add_with_type(section, "mfg", board.mfg.val, board.mfg.type);
+			json_object_add_with_type(section, "pname", board.pname.val, board.pname.type);
+			json_object_add_with_type(section, "serial", board.serial.val, board.serial.type);
+			json_object_add_with_type(section, "pn", board.pn.val, board.pn.type);
+			json_object_add_with_type(section, "file", board.file.val, board.file.type);
+			temp_obj = json_object_new_array();
+			fru_reclist_t *next = board.cust;
+			while (next != NULL) {
+				json_object_add_with_type(temp_obj, NULL, next->rec->data,
+							  typelen2ind(next->rec->typelen));
+				next = next->next;
+			}
+			json_object_object_add(section, "custom", temp_obj);
+			json_object_object_add(json_root, "board", section);
+		}
+		// dump to stdout
+		json_object_to_fd(fileno(stdout), json_root, 0);
```

Comment:
This function doesn't exist in my version of json-c. Probably, you need to demand some minimum version of the library in CMakeLists.txt? Better though, make the code compatible with version 0.11 for compatibility with legacy distributions like Ubuntu 16.04 (xenial).

Also, your patches produce a lot of new warnings for Ubuntu 16.04. That may be the same issue as above, but anyway please make sure that the code builds without warnings, or at least without new ones.

Reply:
I used json-c 0.16-2. I'll double-check the code with Ubuntu 16.04